projects/uvis_gs_fail/training_validation.py

```diff
@@ -0,0 +1,613 @@
+
+    Parameters
+    ----------
+    valid_loader : Torch tensor
+        The validation data to run through the validation loop formatted as a 
+        Torch tensor object.
+
+    Returns 
+    -------
+    val_loss_norm : float
+        The normalized validation loss metric for the current epoch.
+    accuracy : float
+        The accuracy of the model in the current epoch.
+    precision :  float
+        The precision metric of the model for the current epoch.
+    recall :  float
+        The recall metric for the current epoch.
+    fscores : float
+        The fscore for the current epoch.
+    '''
+    # Change model to evaluate mode (deactivates backpropogation)
+    model.eval()
+    
+    # Initialize validation loss and number of correct predictions
+    val_loss = 0
+    correct = 0
+    
+    # Do not calculate gradients for the loop
+    with torch.no_grad():
+        
+        # Loop through batches of validation data
+        for data, target in valid_loader:
+            
+            # Put validation batch on device
+            data = data.float().to(device)
+            target = target.type(torch.LongTensor).to(device)
+            
+            # Calculate output and loss from validation batch
+            output = model(data)
+            val_loss += distance(output, target).item()
+            
+            # Count number of correct predictions
+            pred = output.data.max(1, keepdim=True)[1]
+            correct += pred.eq(target.data.view_as(pred)).sum()
+        
+        # Calculate accuracy
+        accuracy = 100. * correct / len(valid_loader.dataset)
+        
+        # Calculate validation and recall for the epoch
+        prec_global,rec_global,fscore,support = precision_recall_fscore_support\
+                             (target,pred, average = 'micro', zero_division=1.0)
+        
+        precision = prec_global
+        recall = rec_global
+        fscores = fscore
+
+    # Normalize validation loss from one epoch
+    val_loss_norm = val_loss / len(valid_loader)
+    
+    return val_loss_norm, accuracy, precision, recall, fscores
+
+def main(path, metric_dir, data_type = 'None'):
+    '''Load, organize, and format data sets, and train a model on that data.
+
+    Performs the following steps for a model:
+    1. Calls organize_data to get all the processed data and puts them into 
+    tuples for augmented and non-augmented data.
+    2. Uses a boolean to decide wether to use augmented or non-augmented data, 
+    and splits the datasets into x and y datasets for training, and validation 
+    sets.
+    3. Set the hyperparameters, and format the training and validation sets as
+    callable methods.
+    4. Run the training and validation loops for the specified number of epochs
+    and save the state dictionary every 10 epochs.
+    5. Save the training and validation evaluation metrics for each epoch of 
+    training to csv files.
+
+    Parameters
+    -----------
```

Comment:
Use the same definitions for the parameters as listed in other functions (keeps everything consistent).